File: src/main/java/com/pubnub/api/managers/MapperManager.java
package com.pubnub.api.managers;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;
import com.pubnub.api.PubNubException;
import com.pubnub.api.builder.PubNubErrorBuilder;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import retrofit2.Converter;
import retrofit2.converter.gson.GsonConverterFactory;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Iterator;
import java.util.Map;

public class MapperManager {

    @Getter
    private final Gson objectMapper;
    @Getter
    private final Converter.Factory converterFactory;

    private final ObjectMapper jacksonObjectMapper = new ObjectMapper();

    public MapperManager() {
        TypeAdapter<Boolean> booleanAsIntAdapter = getBooleanTypeAdapter();

        this.objectMapper = new GsonBuilder()
                .registerTypeAdapter(Boolean.class, booleanAsIntAdapter)
                .registerTypeAdapter(boolean.class, booleanAsIntAdapter)
                .registerTypeAdapter(JSONObject.class, new JSONObjectAdapter())
                .registerTypeAdapter(JSONArray.class, new JSONArrayAdapter())
                .disableHtmlEscaping()
                .create();
        this.converterFactory = GsonConverterFactory.create(this.getObjectMapper());
    }

    public boolean hasField(JsonElement element, String field) {
        return element.getAsJsonObject().has(field);
    }

    public JsonElement getField(JsonElement element, String field) {
        return element.getAsJsonObject().get(field);
    }

    public Iterator<JsonElement> getArrayIterator(JsonElement element) {
        return element.getAsJsonArray().iterator();
    }

    public Iterator<JsonElement> getArrayIterator(JsonElement element, String field) {
        return element.getAsJsonObject().get(field).getAsJsonArray().iterator();
    }

    public Iterator<Map.Entry<String, JsonElement>> getObjectIterator(JsonElement element) {
        return element.getAsJsonObject().entrySet().iterator();
    }

    public Iterator<Map.Entry<String, JsonElement>> getObjectIterator(JsonElement element, String field) {
        return element.getAsJsonObject().get(field).getAsJsonObject().entrySet().iterator();
    }

    public String elementToString(JsonElement element) {
        return element.getAsString();
    }

    public String elementToString(JsonElement element, String field) {
        return element.getAsJsonObject().get(field).getAsString();
    }

    public int elementToInt(JsonElement element, String field) {
        return element.getAsJsonObject().get(field).getAsInt();
    }

    public boolean isJsonObject(JsonElement element) {
        return element.isJsonObject();
    }

    public JsonObject getAsObject(JsonElement element) {
        return element.getAsJsonObject();
    }

    public boolean getAsBoolean(JsonElement element, String field) {
        return element.getAsJsonObject().get(field).getAsBoolean();
    }

    public void putOnObject(JsonObject element, String key, JsonElement value) {
        element.add(key, value);
    }

    public JsonElement getArrayElement(JsonElement element, int index) {
        return element.getAsJsonArray().get(index);
    }

    public Long elementToLong(JsonElement element) {
        return element.getAsLong();
    }

    public Long elementToLong(JsonElement element, String field) {
        return element.getAsJsonObject().get(field).getAsLong();
    }

    public JsonArray getAsArray(JsonElement element) {
        return element.getAsJsonArray();
    }

    @SuppressWarnings("unchecked")
    public <T> T fromJson(String input, Class<T> clazz) throws PubNubException {
        try {
            return this.objectMapper.fromJson(input, clazz);
        } catch (JsonParseException e) {
            throw PubNubException.builder()
                    .pubnubError(PubNubErrorBuilder.PNERROBJ_PARSING_ERROR)
                    .errormsg(e.getMessage())
                    .cause(e)
                    .build();
        }
    }

    @SuppressWarnings("unchecked")
    public <T> T convertValue(JsonElement input, Class clazz) {
        return (T) this.objectMapper.fromJson(input, clazz);
    }

    @SuppressWarnings("unchecked")
    public <T> T convertValue(Object object, Class clazz) throws PubNubException {
        return (T) fromJson(toJson(object), clazz);
    }

    public JsonElement toJsonTree(Object object) {
        return objectMapper.toJsonTree(object);
    }

    public String toJson(Object input) throws PubNubException {
        try {
            return this.objectMapper.toJson(input);
        } catch (JsonParseException e) {
            throw PubNubException.builder()
                    .pubnubError(PubNubErrorBuilder.PNERROBJ_JSON_ERROR)
                    .errormsg(e.getMessage())
                    .cause(e)
                    .build();
        }
    }

    public String toJsonUsinJackson(Object input) throws PubNubException {
        try {
            return this.jacksonObjectMapper.writeValueAsString(input);
        } catch (JsonProcessingException e) {
            throw PubNubException.builder()
                    .pubnubError(PubNubErrorBuilder.PNERROBJ_JSON_ERROR)
                    .errormsg(e.getMessage())
                    .cause(e)
                    .build();
        }
    }

    @NotNull
    private TypeAdapter<Boolean> getBooleanTypeAdapter() {
        return new TypeAdapter<Boolean>() {
            @Override
            public void write(JsonWriter out, Boolean value) throws IOException {
                if (value == null) {
                    out.nullValue();
                } else {
                    out.value(value);
                }
            }

            @Override
            public Boolean read(JsonReader in) throws IOException {
                JsonToken peek = in.peek();
                switch (peek) {
                    case BOOLEAN:
                        return in.nextBoolean();
                    //case NULL:
                    //    in.nextNull();
                    //    return null;
                    case NUMBER:
                        return in.nextInt() != 0;
                    case STRING:
                        return Boolean.parseBoolean(in.nextString());
                    default:
                        throw new IllegalStateException("Expected BOOLEAN or NUMBER but was " + peek);
                }
            }
        };
    }

    private static class JSONObjectAdapter implements JsonSerializer<JSONObject>, JsonDeserializer<JSONObject> {

        @Override
        public JsonElement serialize(JSONObject src, Type typeOfSrc, JsonSerializationContext context) {
            if (src == null) {
                return null;
            }
            JsonObject jsonObject = new JsonObject();
            Iterator<String> keys = src.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                Object value = src.opt(key);
                JsonElement jsonElement = context.serialize(value, value.getClass());
                jsonObject.add(key, jsonElement);
            }
            return jsonObject;
        }

        @Override
        public JSONObject deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            if (json == null) {
                return null;
            }
            try {
                return new JSONObject(json.toString());
            } catch (JSONException e) {
                e.printStackTrace();
                throw new JsonParseException(e);
            }
        }
    }

    private static class JSONArrayAdapter implements JsonSerializer<JSONArray>, JsonDeserializer<JSONArray> {

        @Override
        public JsonElement serialize(JSONArray src, Type typeOfSrc, JsonSerializationContext context) {
            if (src == null) {
                return null;
            }
            JsonArray jsonArray = new JsonArray();
            for (int i = 0; i < src.length(); i++) {
                Object object = src.opt(i);
                JsonElement jsonElement = context.serialize(object, object.getClass());
                jsonArray.add(jsonElement);
            }
            return jsonArray;
        }

        @Override
        public JSONArray deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            if (json == null) {
                return null;
            }
            try {
                return new JSONArray(json.toString());
            } catch (JSONException e) {
                e.printStackTrace();
                throw new JsonParseException(e);
            }
        }
    }
}

File: src/main/kotlin/com/pubnub/api/managers/MapperManager.kt
package com.pubnub.api.managers

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import com.google.gson.stream.JsonWriter
import com.pubnub.api.PubNubError
import com.pubnub.api.PubNubException
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Converter
import retrofit2.converter.gson.GsonConverterFactory
import java.lang.reflect.Type

class MapperManager {

    private val objectMapper: Gson
    internal val converterFactory: Converter.Factory
    private val jacksonObjectMapper = jacksonObjectMapper()

    init {
        val booleanAsIntAdapter = object : TypeAdapter<Boolean>() {
            override fun write(out: JsonWriter?, value: Boolean?) {
                if (value == null) {
                    out?.nullValue()
                } else {
                    out?.value(value)
                }
            }

            override fun read(_in: JsonReader): Boolean? {
                val peek: JsonToken = _in.peek()
                return when (peek) {
                    JsonToken.BOOLEAN -> _in.nextBoolean()
                    JsonToken.NUMBER -> _in.nextInt() != 0
                    JsonToken.STRING -> java.lang.Boolean.parseBoolean(_in.nextString())
                    else -> throw IllegalStateException("Expected BOOLEAN or NUMBER but was $peek")
                }
            }
        }

        objectMapper = GsonBuilder()
            .registerTypeAdapter(Boolean::class.javaObjectType, booleanAsIntAdapter)
            .registerTypeAdapter(Boolean::class.javaPrimitiveType, booleanAsIntAdapter)
            .registerTypeAdapter(Boolean::class.java, booleanAsIntAdapter)
            .registerTypeAdapter(JSONObject::class.java, JSONObjectAdapter())
            .registerTypeAdapter(JSONArray::class.java, JSONArrayAdapter())
            .create()
        converterFactory = GsonConverterFactory.create(objectMapper)
    }

    @Throws(PubNubException::class)
    fun toJsonUsingJackson(input: Any): String {
        return try {
            this.jacksonObjectMapper.writeValueAsString(input)
        } catch (e: JsonProcessingException) {
            throw PubNubException(PubNubError.JSON_ERROR).copy(errorMessage = e.message ?: PubNubError.JSON_ERROR.message)
        }
    }

    fun hasField(element: JsonElement, field: String) = element.asJsonObject.has(field)

    fun getField(element: JsonElement?, field: String): JsonElement? {
        if (element?.isJsonObject!!) {
            return element.asJsonObject.get(field)
        }
        return null
    }

    fun getArrayIterator(element: JsonElement?) = element?.asJsonArray?.iterator()

    fun getArrayIterator(element: JsonElement, field: String) = element.asJsonObject.get(field).asJsonArray.iterator()

    fun getObjectIterator(element: JsonElement) = element.asJsonObject.entrySet().iterator()

    fun getObjectIterator(element: JsonElement, field: String) =
        element.asJsonObject.get(field).asJsonObject.entrySet().iterator()

    fun elementToString(element: JsonElement?) = element?.asString

    fun elementToString(element: JsonElement?, field: String) = element?.asJsonObject?.get(field)?.asString

    fun elementToInt(element: JsonElement, field: String) = element.asJsonObject.get(field).asInt

    fun isJsonObject(element: JsonElement) = element.isJsonObject

    fun getAsObject(element: JsonElement) = element.asJsonObject

    fun getAsBoolean(element: JsonElement, field: String) = element.asJsonObject.get(field)?.asBoolean
        .run { this != null }

    fun putOnObject(element: JsonObject, key: String, value: JsonElement) = element.add(key, value)

    fun getArrayElement(element: JsonElement, index: Int) = element.asJsonArray.get(index)

    fun elementToLong(element: JsonElement) = element.asLong

    fun elementToLong(element: JsonElement, field: String) = element.asJsonObject.get(field).asLong

    fun getAsArray(element: JsonElement) = element.asJsonArray

    fun <T> fromJson(input: String?, clazz: Class<T>): T {
        return try {
            this.objectMapper.fromJson<T>(input, clazz)
        } catch (e: JsonParseException) {
            throw PubNubException(
                    pubnubError = PubNubError.PARSING_ERROR,
                    errorMessage = e.message
            )
        }
    }

    fun <T> fromJson(input: String?, typeOfT: Type): T {
        return try {
            this.objectMapper.fromJson<T>(input, typeOfT)
        } catch (e: JsonParseException) {
            throw PubNubException(
                    pubnubError = PubNubError.PARSING_ERROR,
                    errorMessage = e.message
            )
        }
    }

    fun <T> convertValue(input: JsonElement?, clazz: Class<T>): T {
        return this.objectMapper.fromJson(input, clazz) as T
    }

    fun <T> convertValue(o: Any?, clazz: Class<T>?): T {
        return this.objectMapper.fromJson(toJson(o), clazz) as T
    }

    fun toJson(input: Any?): String {
        return try {
            this.objectMapper.toJson(input)
        } catch (e: JsonParseException) {
            throw PubNubException(
                    pubnubError = PubNubError.JSON_ERROR,
                    errorMessage = e.message
            )
        }
    }

    private class JSONObjectAdapter : JsonSerializer<JSONObject?>, JsonDeserializer<JSONObject?> {
        override fun serialize(
            src: JSONObject?,
            typeOfSrc: Type?,
            context: JsonSerializationContext
        ): JsonElement? {
            if (src == null) {
                return null
            }
            val jsonObject = JsonObject()
            val keys: Iterator<String> = src.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val value: Any = src.opt(key)
                val jsonElement = context.serialize(value, value.javaClass)
                jsonObject.add(key, jsonElement)
            }
            return jsonObject
        }

        override fun deserialize(
            json: JsonElement?,
            typeOfT: Type?,
            context: JsonDeserializationContext?
        ): JSONObject? {
            return if (json == null) {
                null
            } else try {
                JSONObject(json.toString())
            } catch (e: JSONException) {
                e.printStackTrace()
                throw JsonParseException(e)
            }
        }
    }

    private class JSONArrayAdapter : JsonSerializer<JSONArray?>, JsonDeserializer<JSONArray?> {
        override fun serialize(
            src: JSONArray?,
            typeOfSrc: Type?,
            context: JsonSerializationContext
        ): JsonElement? {
            if (src == null) {
                return null
            }
            val jsonArray = JsonArray()
            for (i in 0 until src.length()) {
                val `object`: Any = src.opt(i)
                val jsonElement = context.serialize(`object`, `object`.javaClass)
                jsonArray.add(jsonElement)
            }
            return jsonArray
        }

        override fun deserialize(
            json: JsonElement?,
            typeOfT: Type?,
            context: JsonDeserializationContext?
        ): JSONArray? {
            return if (json == null) {
                null
            } else try {
                JSONArray(json.toString())
            } catch (e: JSONException) {
                e.printStackTrace()
                throw JsonParseException(e)
            }
        }
    }
}
